fix(admin/performance): Fix default time bug in Schedule policy

rafpaf opened the pull request:
Closes #45916

### Description

On master there is the following bug: if you switch to a weekly or monthly schedule strategy, but leave the time at the default value of 8am, the schedule that will get saved in the BE will use the _wildcard_ for the hour (rather than the integer 8), resulting in many more cache invalidations than you wanted.

Here's an illustration of the bug and the fix: https://www.loom.com/share/7065bd50b8a14f9ea4071af7fe327631?sid=e82d2ab7-993b-4f41-99c1-4bc4ef1e434f

### How to verify

Describe the steps to verify that the changes are working as expected.

1. Admin / Performance
2. Take a database that does not use the schedule strategy and change it to the schedule strategy. Choose weekly. Don't change the time.
3. On master, you'll see in DevTools / Network that there is no integer 8 in the cron expression sent to the BE. On this branch, you'll see an 8.

### Checklist

[x] Tests added here (this required bringing in some helpers from a downstream branch)

rafpaf commented:
* **#45915** <a href="https://app.graphite.dev/github/pr/metabase/metabase/45915?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rafpaf and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

iethree reviewed (changes requested):
Looks good, but we shouldn't merge this without the e2e tests in this branch. It's ok to take some more time to package it up nicely so we can be sure it stays fixed.

rafpaf committed:
fix(admin/performance): Fix bug where leaving the hour at its default value (8am) would result in a cron expression with a wildcard for the hour

rafpaf committed:
fix(admin/performance): Fix bug where leaving the hour at its default value (8am) would result in a cron expression with a wildcard for the hour

rafpaf committed:
reset change that unblocked my cypress

rafpaf committed:
reset change that unblocked my cypress

rafpaf commented on e2e/test/scenarios/admin/performance/performance.cy.spec.ts:
```diff
@@ -1,46 +1,17 @@
 import { describeEE, restore, setTokenFeatures } from "e2e/support/helpers";
 
```
I've moved these helper functions to a separate file and changed them a bit, so these tests need to change a bit

rafpaf commented on e2e/test/scenarios/admin/performance/schedule.cy.spec.ts:
These are the main e2e tests added in this Pr

rafpaf commented on e2e/test/scenarios/dashboard/dashboard.cy.spec.js:
A small refactoring of the sidebar code to facilitate testing required a change to this file

rafpaf commented on frontend/src/metabase/components/Schedule/components.tsx:
```diff
@@ -34,18 +37,54 @@ export type SelectFrameProps = {
   frames?: { label: string; value: ScheduleFrameType }[];
 };
 
```
I moved this function here to be next to the other `Select*` components

uladzimirdev committed:
perf/fix schedule 8am bug 2 (#46062)

### Description

Adds custom resolvers for svg files which couldn't be loaded by esbuild for some reason

![image (11)](https://github.com/user-attachments/assets/5a95bb22-9088-4ef4-ab4f-f9897a3247fe)


### How to verify

ci is green

rafpaf committed:
comment out some tests

rafpaf committed:
Fix e2e tests

rafpaf committed:
restore tests

rafpaf committed:
Add some fixmes

rafpaf committed:
Improve test

rafpaf commented on e2e/support/config.js:
```diff
@@ -21,6 +23,23 @@ const targetVersion = process.env["CROSS_VERSION_TARGET"];
+      };
+    });
+  },
+};
```
This fix was added by @uladzimirdev due to Cypress build issue

rafpaf commented on e2e/test/scenarios/admin/performance/schedule.cy.spec.ts:
```diff
@@ -0,0 +1,160 @@
+            throw new Error(
+              `Unexpected schedule frequency: ${schedule.frequency}`,
+            );
+        }
```
We can't use ts-pattern here because this needs to get backported :-(

rafpaf committed:
fix test

rafpaf committed:
Fix tests

rafpaf committed:
Add note

rafpaf committed:
Don't use ts pattern since this is a backport

iethree commented on e2e/test/scenarios/admin/performance/schedule.cy.spec.ts:
```diff
@@ -0,0 +1,163 @@
+
+    [{ frequency: "daily" }, "0 0 8 * * ?"], // 8 am is the default time for a daily schedule
+
+    [{ frequency: "hourly" }, "0 0 * * * ?"], // Check that switching back to hourly works
+
+    [{ frequency: "daily", time: "9:00" }, "0 0 9 * * ?"], // AM is the default choice between AM and PM
+    [{ frequency: "daily", time: "9:00", amPm: "PM" }, "0 0 21 * * ?"],
+    [{ frequency: "daily", time: "12:00", amPm: "AM" }, "0 0 0 * * ?"],
+
+    [ { frequency: "weekly", weekday: "Monday", time: "12:00", amPm: "AM" }, "0 0 0 ? * 2" ],
+    [ { frequency: "weekly", weekday: "Monday", time: "1:00", amPm: "AM" }, "0 0 1 ? * 2" ],
+
+    [ { frequency: "weekly", weekday: "Tuesday", time: "2:00", amPm: "AM" }, "0 0 2 ? * 3" ],
+    [ { frequency: "weekly", weekday: "Tuesday", time: "3:00", amPm: "AM" }, "0 0 3 ? * 3" ],
+
+    [ { frequency: "weekly", weekday: "Wednesday", time: "4:00", amPm: "AM" }, "0 0 4 ? * 4" ],
+    [ { frequency: "weekly", weekday: "Wednesday", time: "5:00", amPm: "AM" }, "0 0 5 ? * 4" ],
+
+    [ { frequency: "weekly", weekday: "Thursday", time: "6:00", amPm: "AM" }, "0 0 6 ? * 5" ],
+    [ { frequency: "weekly", weekday: "Thursday", time: "7:00", amPm: "AM" }, "0 0 7 ? * 5" ],
+
+    [ { frequency: "weekly", weekday: "Friday", time: "8:00", amPm: "AM" }, "0 0 8 ? * 6" ],
+    [ { frequency: "weekly", weekday: "Friday", time: "9:00", amPm: "AM" }, "0 0 9 ? * 6" ],
+
+    [ { frequency: "weekly", weekday: "Saturday", time: "10:00", amPm: "AM" }, "0 0 10 ? * 7" ],
+    [ { frequency: "weekly", weekday: "Saturday", time: "11:00", amPm: "AM" }, "0 0 11 ? * 7" ],
+
+    [ { frequency: "weekly", weekday: "Sunday", time: "12:00", amPm: "PM" }, "0 0 12 ? * 1" ],
+    [ { frequency: "weekly", weekday: "Sunday", time: "1:00", amPm: "PM" }, "0 0 13 ? * 1" ],
+
+    [ { frequency: "monthly", frame: "first", weekdayOfMonth: "Sunday", time: "12:00", amPm: "AM" }, "0 0 0 ? * 1#1" ],
+    [ { frequency: "monthly", frame: "first", weekdayOfMonth: "Monday", time: "2:00", amPm: "AM" }, "0 0 2 ? * 2#1" ],
+
+    [ { frequency: "monthly", frame: "last", weekdayOfMonth: "Tuesday", time: "12:00", amPm: "AM" }, "0 0 0 ? * 3L" ],
+    [ { frequency: "monthly", frame: "last", weekdayOfMonth: "Wednesday", time: "12:00", amPm: "AM" }, "0 0 0 ? * 4L" ],
+
+    [ { frequency: "monthly", frame: "15th", time: "12:00", amPm: "AM" }, "0 0 0 15 * ?" ],
+    [ { frequency: "monthly", frame: "15th", time: "11:00", amPm: "PM" }, "0 0 23 15 * ?" ],
+
+  ];
```
I'll let you be the judge but keep in mind that with e2e tests:
- they take a long time
- so we try to use them sparingly to cover complex UI cases
- that can't easily be covered by much faster unit tests

You know this code better than I do, but it looks to me that we're exercising most of this functionality multiple times, and some of the more granular testing can be handled in unit tests. (e.g. it's probably not worth it to have a test for every day of the week, probably a single hourly/daily/weekly/monthly test is enough for e2e, and you can drill down in more detail in unit tests)

iethree reviewed (approved):
Looks good! thanks for adding the tests 💪 

take a look at the note re unnecessary test cases.
